client.go

```diff
@@ -153,29 +158,73 @@ func (r remoteStdin) Close() error {
 }
 
 type pipe struct {
-	r *io.PipeReader
-	w *io.PipeWriter
+	r   *io.PipeReader
+	w   *io.PipeWriter
+	d   chan []byte
+	e   chan error
+	buf []byte
 }
 
 func newPipe() pipe {
 	pr, pw := io.Pipe()
 	return pipe{
-		r: pr,
-		w: pw,
+		r:   pr,
+		w:   pw,
+		d:   make(chan []byte),
+		e:   make(chan error),
+		buf: make([]byte, maxMessageSize),
 	}
 }
 
-func (r *remoteProcess) listen(ctx context.Context) {
-	defer close(r.done)
-	defer func() {
-		r.closeErr = r.conn.Close(websocket.StatusNormalClosure, "normal closure")
+// writeCtx writes data to the pipe, or returns if the context is canceled.
+func (p *pipe) writeCtx(ctx context.Context, data []byte) error {
+	// actually do the copy on another goroutine so that we can return if context
+	// is canceled
+	go func() {
+		var err error
+		select {
+		case <-ctx.Done():
+			return
+		case body := <-p.d:
+			_, err = io.CopyBuffer(p.w, bytes.NewReader(body), p.buf)
```

Comment:
Was this because this would hang, preventing the other function from closing and running the "close" on the defers? Makes sense.

Do we need the channels to be on the pipe though? The channels are only used in `writeCtx`, it would be nice if the channels don't exist outside of this function. Although I understand if for performance reasons you do not want to reallocate them each time.

Just a note, overall looks good.

Reply:
> Was this because this would hang, preventing the other function from closing and running the "close" on the defers? Makes sense.

Yeah, exactly.  If you stopped reading to close the process, you kind of have no way of knowing whether you've got everything, and if not, then it would hang on close.

> Do we need the channels to be on the pipe though? The channels are only used in writeCtx, it would be nice if the channels don't exist outside of this function. Although I understand if for performance reasons you do not want to reallocate them each time.


Yeah, that was my thought: allocating two channels (which I think always involve the heap) for every write seems like we'd be thrashing GC when we could avoid it.  Same story with the buffer.